tests/test_io.py

```diff
@@ -0,0 +1,72 @@
+import os
+import tempfile
+from typing import Any
+
+from pandas import DataFrame
+from typing_extensions import assert_type
+
+from tests import check
+
+from pandas.io.pickle import (
+    read_pickle,
+    to_pickle,
+)
+
+DF = DataFrame({"a": [1, 2, 3], "b": [0.0, 0.0, 0.0]})
+
+
+def test_pickle():
+    with tempfile.NamedTemporaryFile(delete=False) as file:
+        check(assert_type(DF.to_pickle(file), None), type(None))
+        file.seek(0)
+        check(assert_type(read_pickle(file.name), Any), DataFrame)
+        file.close()
+        check(assert_type(read_pickle(file.name), Any), DataFrame)
+        os.unlink(file.name)
+
+    with tempfile.NamedTemporaryFile(delete=False) as file:
+        check(assert_type(to_pickle(DF, file), None), type(None))
```

Comment:
It is listed in pandas/io/api.py so I assume this makes it public even it not on the docs.

Reply:
> It is listed in pandas/io/api.py so I assume this makes it public even it not on the docs.

There's a few schools of thought here:
1. We trim down `pandas-stubs` so that it only type checks what is in the documented public API
2. If a function or class is "public" in the sense that it does not begin with an underscore but is not documented, we create a stub for it and test it.
3.  If a function or class is "public" in the sense that it does not begin with an underscore and is not documented, we are agnostic on creating a stub or testing that stub.

So far, @twoertwein and I have been leaning towards (1).  With `pandas.io.api.to_pickle()`, you are proposing (2).  

@twoertwein  What are your thoughts?

Reply:
I think it is more of what is the "API". To me it is something like

* Documented as part of the docs
* Imported into an `api.py` file or the top level `__init__.py file.

I think 2 is too broad because there hasn't been enough effort in pandas to `_` modules, classes and methods. 

In short, if it seems to be part of an API, then it is reasonable to include it. 

A related point is documenting public methods of classes that appear parrtially in the docs, something like `Klass.method(arg1, arg2)`. Should public of `Klass` be documented, or just `Klass.__init__` and `Klass.method`

Reply:
My goal is that everything that is meant to be public (which is often unclear) is documented and in pandas-stubs. Personally, I think the best way is to remove any symbol from the stubs that is not meant to be public.

- If it seems reasonable that it is meant to be public, it might be a better user experience, if we first open an issue at pandas before potentially removing it.
- If we remove too much, we get user feedback and can then create an issue at pandas.


I believe typeshed uses `# undocumented` to indicate which parts of their stubs are technically not documented. if we want to keep more in here, we could follow that approach. Luckily we have rather good connections to pandas :) so we might just open an issue there :)

Reply:
There are some grey zones: a private super class but the inherited methods are public in a public child class: I would keep the parent class (in the long-term, I would like if pandas-stubs aligns with pandas), define `__all__` but exclude the class from it.